### PR TITLE
Add docs for request timeouts middleware

### LINKED ISSUE #28934: Add docs for request timeouts middleware
.NET 8 has a new request timeouts middleware, which allows users to set request timeouts for individual endpoints, controllers, or dynamically per request. We should document it.

Some notes that may be useful...

To apply request timeouts, first add the request timeout services:

 ```csharp
 builder.Services.AddRequestTimeouts();
 ```

 You can then apply request timeouts using middleware by calling `UseRequestTimeouts()`:

 ```csharp
 app.UseRequestTimeouts();
 ```

 To apply request timeouts to a particular endpoint call `WithRequestTimeout(timeout)` or add `[RequestTimeout(timeout)]` to the controller or action.

 The timeouts are cooperative; when they expire, the `HttpContext.RequestAborted` cancellation token will trigger but the request will not be forcibly aborted. It's up to the application to monitor the token and decide how to end the request processing.

---

More info at: 
- Issue: https://github.com/dotnet/aspnetcore/issues/45732
- PR: https://github.com/dotnet/aspnetcore/pull/46115

***EDIT by @Rick-Anderson to add the following metadata***

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: 9898482c-fa54-7ce2-1716-d14fbe03990d
* Version Independent ID: 3926c686-0ccc-f9fe-ca52-0134612a1623
* Content: [ASP.NET Core Middleware](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/middleware/?view=aspnetcore-7.0)
* Content Source: [aspnetcore/fundamentals/middleware/index.md](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/fundamentals/middleware/index.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-fundamentals**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

---
[Associated WorkItem - 85532](https://dev.azure.com/msft-skilling/Content/_workitems/edit/85532)

### PR DESCRIPTION
**Pull request by tdykstra**
Fixes #28934 

API ref links will be added when the Preview 3 API ref is published.

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [aspnetcore/fundamentals/middleware/index.md](https://github.com/dotnet/AspNetCore.Docs/blob/66e84e18089e30d7fe177a40eeb71cee7a20a117/aspnetcore/fundamentals/middleware/index.md) | [ASP.NET Core Middleware](https://review.learn.microsoft.com/en-us/aspnet/core/fundamentals/middleware/index?branch=pr-en-us-29068) |
| [aspnetcore/fundamentals/minimal-apis.md](https://github.com/dotnet/AspNetCore.Docs/blob/66e84e18089e30d7fe177a40eeb71cee7a20a117/aspnetcore/fundamentals/minimal-apis.md) | [Differences between minimal APIs and APIs with controllers](https://review.learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis?branch=pr-en-us-29068) |
| [aspnetcore/performance/overview.md](https://github.com/dotnet/AspNetCore.Docs/blob/66e84e18089e30d7fe177a40eeb71cee7a20a117/aspnetcore/performance/overview.md) | [ASP.NET Core performance](https://review.learn.microsoft.com/en-us/aspnet/core/performance/overview?branch=pr-en-us-29068) |
| [aspnetcore/performance/timeouts.md](https://github.com/dotnet/AspNetCore.Docs/blob/66e84e18089e30d7fe177a40eeb71cee7a20a117/aspnetcore/performance/timeouts.md) | [aspnetcore/performance/timeouts](https://review.learn.microsoft.com/en-us/aspnet/core/performance/timeouts?branch=pr-en-us-29068) |

### CODE REVIEW

**Comment by tdykstra** [comment]
@Kahbazi please review

File: aspnetcore/performance/timeouts/samples/8.x/Program.cs [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    await Task.Delay(TimeSpan.FromSeconds(2));
+
+    if (context.RequestAborted.IsCancellationRequested)
+    {
+        var timeoutFeature = context.Features.Get<IHttpRequestTimeoutFeature>();
+        timeoutFeature?.DisableTimeout();
+    }
+
+    return Results.Content("No timeout!", "text/plain");
+});
```
Comment (Kahbazi) [inline aspnetcore/performance/timeouts/samples/8.x/Program.cs:151]: I think this sample shows better that how `DisableTimeout` would work.

```suggestion
app.MapGet("/canceltimeout", async (HttpContext context) => {

    var timeoutFeature = context.Features.Get<IHttpRequestTimeoutFeature>();
    timeoutFeature?.DisableTimeout();

    await Task.Delay(TimeSpan.FromSeconds(2));
    
    if (context.RequestAborted.IsCancellationRequested) // is false
    {

    }

    return Results.Content("No timeout!", "text/plain");
}).WithRequestTimeout(TimeSpan.FromSeconds(1));
```

File: aspnetcore/performance/timeouts.md [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+Add the middleware to the request processing pipeline by calling `UseRequestTimeouts`>.
+
+> [!NOTE]
+> * In apps that explicitly call `UseRouting`, `UseRequestTimeouts` must be called after `UseRouting`.
+
+Adding the middleware to the app doesn't automatically start triggering timeouts. Timeout limits have to be explicitly configured.
+
+## Configure one endpoint or page
+
+For minimal API apps, configure an endpoint to time out by calling `WithRequestTimeouts`, or by applying the `[RequestTimeouts]` attribute, as shown in the following example:
```
Comment (Kahbazi) [inline aspnetcore/performance/timeouts.md:37]:
```suggestion
For minimal API apps, configure an endpoint to time out by calling `WithRequestTimeout`, or by applying the `[RequestTimeout]` attribute, as shown in the following example:
```

File: aspnetcore/performance/timeouts.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+The `RequestTimeoutPolicy` has a `WriteTimeoutResponse` property that can be used to customize the response when a timeout is triggered.
+
+:::code language="csharp" source="~/performance/timeouts/samples/8.x/Program.cs" id="definepolicies2" highlight="8-11":::
+
+:::code language="csharp" source="~/performance/timeouts/samples/8.x/Program.cs" id="usepolicy2" highlight="4,7":::
+
+## Disable timeouts
+
+To disable all timeouts including the default global timeout, use the `[DisableRequestTimeout]` attribute:
```
[the remark this answers is not in the document]
Comment (tdykstra) [reply aspnetcore/performance/timeouts.md]: Thanks, I added that.

**Comment by tdykstra** [comment]
Thanks for the review, @Kahbazi. Are my comments about middleware order accurate and sufficient (just saying that `UseRequestTimeouts` goes after `UseRouting`)?

File: aspnetcore/performance/timeouts.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+To disable all timeouts including the default global timeout, use the `[DisableRequestTimeout]` attribute or the `DisableRequestTimeout` extension method:
+
+:::code language="csharp" source="~/performance/timeouts/samples/8.x/Program.cs" id="disablebyattr" highlight="1":::
+
+:::code language="csharp" source="~/performance/timeouts/samples/8.x/Program.cs" id="disablebyext" highlight="5":::
+
+## Cancel a timeout
+
+To cancel a timeout that has already been triggered, use the `DisableTimeout` method on `IHttpRequestTimeoutFeature`.
```
Comment (Tratcher) [inline aspnetcore/performance/timeouts.md:93]:
```suggestion
To cancel a timeout that has already been started, use the `DisableTimeout` method on `IHttpRequestTimeoutFeature`. Timeouts cannot be canceled after they've expired.
```

**Review by Tratcher — APPROVED** [review]

**Comment by Kahbazi** [comment]
> Thanks for the review, @Kahbazi. Are my comments about middleware order accurate and sufficient (just saying that `UseRequestTimeouts` goes after `UseRouting`)?

Yeah, it looks good.

**Pull request merged**